subsys/net/lib/rest_client/src/rest_client.c

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <string.h>
+#include <zephyr.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <assert.h>
+
+#if defined(CONFIG_POSIX_API)
+#include <posix/arpa/inet.h>
+#include <posix/unistd.h>
+#include <posix/netdb.h>
+#include <posix/sys/socket.h>
+#else
+#include <net/socket.h>
+#endif
+#include <net/tls_credentials.h>
+#include <net/http_client.h>
+#include <net/http_parser.h>
+
+#include <logging/log.h>
+
+#include <net/rest_client.h>
+
+LOG_MODULE_REGISTER(rest_client, CONFIG_REST_CLIENT_LOG_LEVEL);
+
+#define HTTP_PROTOCOL "HTTP/1.1"
+
+static void rest_client_http_response_cb(struct http_response *rsp,
+					  enum http_final_call final_data,
+					  void *user_data)
+{
+	struct rest_client_req_resp_context *rest_ctx = NULL;
+
+	if (user_data) {
+		rest_ctx = (struct rest_client_req_resp_context *)user_data;
+	}
+
+	/* If the entire HTTP response is not received in a single "recv" call
+	 * then this could be called multiple times, with a different value in
+	 * rsp->body_start. Only set rest_ctx->response once, the first time,
+	 * which will be the start of the body.
+	 */
+	if (rest_ctx && !rest_ctx->response && rsp->body_found && rsp->body_start) {
+		rest_ctx->response = rsp->body_start;
+	}
+	rest_ctx->total_response_len += rsp->data_len;
```

Comment:
This will cause a crash if rest_ctx is NULL. Is it necessary for the function to do something if user_data is NULL, or could it just return immediately in that case?

Reply:
Modified code for this.